File: library/core/src/main/java/com/lacker/micros/domain/entity/EntityImpl.java
package com.lacker.micros.domain.entity;

import com.lacker.micros.utils.id.SequenceIdGenerator;
import java.util.Objects;

public abstract class EntityImpl implements Entity {

    private Long id;

    @Override
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || this.id == null || getClass() != o.getClass()) {
            return false;
        }
        EntityImpl entity = (EntityImpl) o;
        return Objects.equals(id, entity.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        EntityImpl clone = (EntityImpl) super.clone();
        clone.id = SequenceIdGenerator.generateId();
        return clone;
    }
}

File: library/core/src/main/kotlin/com/mlacker/micros/domain/entity/EntityImpl.kt
package com.mlacker.micros.domain.entity

import com.mlacker.micros.domain.annotation.NoArg
import com.mlacker.micros.utils.id.SequenceIdGenerator.Companion.generateId
import java.util.*

@NoArg
abstract class EntityImpl (id: Long) : Entity, Cloneable {
    final override var id: Long = id
        private set

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        return id == (other as Entity).id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    override fun clone(): Any {
        val clone = super.clone() as EntityImpl
        clone.id = generateId()
        return clone
    }
}